scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordered_serialization/providers/OptionOrderedBuf.scala

```diff
@@ -0,0 +1,170 @@
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+package com.twitter.scalding.macros.impl.ordered_serialization.providers
+
+import scala.language.experimental.macros
+import scala.reflect.macros.Context
+
+import com.twitter.scalding._
+import com.twitter.scalding.macros.impl.ordered_serialization.{ CompileTimeLengthTypes, ProductLike, TreeOrderedBuf }
+import CompileTimeLengthTypes._
+import com.twitter.scalding.serialization.OrderedSerialization
+
+object OptionOrderedBuf {
+  def dispatch(c: Context)(buildDispatcher: => PartialFunction[c.Type, TreeOrderedBuf[c.type]]): PartialFunction[c.Type, TreeOrderedBuf[c.type]] = {
+    case tpe if tpe.erasure =:= c.universe.typeOf[Option[Any]] => OptionOrderedBuf(c)(buildDispatcher, tpe)
+  }
+
+  def apply(c: Context)(buildDispatcher: => PartialFunction[c.Type, TreeOrderedBuf[c.type]], outerType: c.Type): TreeOrderedBuf[c.type] = {
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+    val dispatcher = buildDispatcher
+
+    val innerType = outerType.asInstanceOf[TypeRefApi].args.head
+    val innerBuf: TreeOrderedBuf[c.type] = dispatcher(innerType)
+
+    def genBinaryCompare(inputStreamA: TermName, inputStreamB: TermName) = {
+      val valueOfA = freshT("valueOfA")
+      val valueOfB = freshT("valueOfB")
+      val tmpHolder = freshT("tmpHolder")
+      q"""
+        val $valueOfA = $inputStreamA.readByte
+        val $valueOfB = $inputStreamB.readByte
+        val $tmpHolder = _root_.java.lang.Byte.compare($valueOfA, $valueOfB)
+        if($tmpHolder != 0 || $valueOfA == (0: Byte)) {
+          $tmpHolder
+        } else {
+          ${innerBuf.compareBinary(inputStreamA, inputStreamB)}
+        }
+      """
+    }
+
+    def genHashFn(element: TermName) = {
+      val innerValue = freshT("innerValue")
+
+      q"""
+        if($element.isEmpty)
+          0
+        else {
+          val $innerValue = $element.get
+          ${innerBuf.hash(innerValue)}
+        }
+      """
+    }
+
+    def genGetFn(inputStreamA: TermName) = {
+      val tmpGetHolder = freshT("tmpGetHolder")
+      q"""
+        val $tmpGetHolder = $inputStreamA.readByte
+        if($tmpGetHolder == 0) {
+          None
+        } else {
+          Some(${innerBuf.get(inputStreamA)})
+        }
+      """
+    }
+
+    def genPutFn(inputStream: TermName, element: TermName) = {
+      val tmpPutVal = freshT("tmpPutVal")
+      val innerValue = freshT("innerValue")
+
+      q"""
+        val $tmpPutVal: _root_.scala.Byte = if($element.isDefined) 1 else 0
+        $inputStream.writeByte($tmpPutVal)
+        if($tmpPutVal == 1) {
+        val $innerValue = $element.get
```

Comment:
Without auto indenting the indentation in some of the quasiquotes is questionable. I've most sane, but... any you see not